Signing request (#92)
cs3org/OCM-API: README.md

When a sending server allows sharing to any internet-hosted receiving server, then discovery can happen from the sharee address, using the `/.well-known/ocm` (or `/ocm-provider`, for backwards compatibility) URL that receiving servers SHOULD provide according to this [specification](https://cs3org.github.io/OCM-API/docs.html?branch=develop&repo=OCM-API&user=cs3org#/paths/~1.well-known~1ocm/get).

> If the identity of an incoming request needs to be confirmed, the discovery data SHOULD contain a public key. Each incoming request that requires to origin from an authenticated source must be signed in its headers using the related private key.

Reviewer: ```suggestion
To ease the process of confirming the identity of a remote party, the discovery data MAY contain a public key: each incoming request that requires to origin from an authenticated source MUST be signed in its headers using the private key of that source, whose public key MUST be exposed in its discovery data.
```

To fill the gap between users knowning other peers' email addresses of the form `user@provider.org`, and the actual cloud storage endpoints being in the form `https://my-cloud-storage.provider.org`, a further discovery mechanism SHOULD be provided by implementations that wish to allow sending shares to any receiver, based on DNS `SRV` Service Records.

* A provider SHOULD ensure that a `type=SRV` DNS query to `_ocm._tcp.provider.org` resolves to e.g. `service = 10 10 443 my-cloud-storage.provider.org`
* When requested to discover the EFSS endpoint for `user@provider.org`, implementations SHOULD query the corresponding `_ocm._tcp.domain` DNS record, e.g. `_ocm._tcp.provider.org`, and subsequently make a HTTP GET request to the host returned by that DNS query, followed by the `/.well-known/ocm` URL path.

> 
> 

Reviewer: ```suggestion
```

The concept is to add unique metadata and sign them using a private/public key pair.
The location of the public key used to verify the signature will confirm the origin of the request.

Signature does not affect the data of the request, it only adds headers to it:
```
 {
     "(request-target)": "post /path",
     "content-length": 380,
     "date": "Mon, 08 Jul 2024 14:16:20 GMT",
     "digest": "SHA-256=U7gNVUQiixe5BRbp4Tg0xCZMTcSWXXUZI2\\/xtHM40S0=",
     "host": "hostname.of.the.recipient",
     "Signature": "keyId=\"https://author.hostname/key\",algorithm=\"ras-sha256\",headers=\"content-length date digest host\",signature=\"DzN12OCS1rsA[...]o0VmxjQooRo6HHabg==\""
 }
```

- 'content-length' is the total length of the payload of the request
- 'date' is the datetime the request have been initiated
- 'digest' is a checksum of the payload of the request
- 'host' is the hostname of the recipient of the request (remote when signing outgoing request, local on incoming request)
- 'Signature' contains the signature generated using the private key, and metadata:

>   * 'keyId' is a unique id, formatted as an url. hostname is used to retrieve the public key via custom discovery

Reviewer 2: from the specs this important part is missing here:

> [...] The hostname set the origin of the request and **MUST** be identical to the current discovery endpoint

Since there is no way to guarantee that the sharee OCM provider will actually enforce the MFA requirement, it is up to the sharer OCM provider to establish a trust with the OCM sharee provider such that it is reasonable to assume that the sharee OCM provider will honor the MFA requirement. This establishment of trust will inevitably be implementation dependent, and can be done for example using a pre approved allow list of trusted OCM providers. The procedure of establishing trust is out of scope for this specification: a mechanism similar to the [ScienceMesh](https://sciencemesh.io) integration for the [Invite](#invite) capability may be envisaged.

## Signing request

> A request is signed by adding the signature in the headers. The sender also need to expose the public key used to generate the signature. The receiver can then validate the signature and therefore the origin of the request.

Reviewer: ```suggestion
A request is signed by adding the signature in the headers. The sender also needs to expose the public key used to generate the signature. The receiver can then validate the signature and therefore the origin of the request.
```

> To helps debugging, it is made mandatory to also add all properties used in the signature as headers, even if they can be easily re-generated from the payload.  

Reviewer: ```suggestion
To help debugging, it is recommended to also add all properties used in the signature as headers, even if they can easily be re-generated from the payload.  
```

Note: Signed requests prove the identity of the sender but does not encrypt nor affect its payload.

Here is an example of headers needed to sign a request.  

```
  {
    "(request-target)": "post /path",
    "content-length": 380,
    "date": "Mon, 08 Jul 2024 14:16:20 GMT",
    "digest": "SHA-256=U7gNVUQiixe5BRbp4Tg0xCZMTcSWXXUZI2\\/xtHM40S0=",
    "host": "hostname.of.the.recipient",

>     "Signature": "keyId=\"https://author.hostname/key\",algorithm=\"ras-sha256\",headers=\"content-length date digest host\",signature=\"DzN12OCS1rsA[...]o0VmxjQooRo6HHabg==\""

Reviewer 3: Shouldn't headers be `"(request-target) content-length date digest host"`? (see also [line 135](https://github.com/cs3org/OCM-API/pull/92/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R135)).
And you have a typo I think, `algorithm` should be `"rsa-sha256"`

Reviewer 3: These are details about [httpsig](https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12), we can fix this in a follow-up.
I also noticed that node refuses to add `(request-target)` as a header (saying it is malformed), so in my implementation I changed it to `request-target`.

  }
```

- '(request-target)' contains the reached endpoint and the used method,
- 'content-length' is the total length of the payload of the request,

> - 'date' is the date and time on which the request have been emitted,

Reviewer: ```suggestion
- 'date' is the date and time when the request has been sent,
```

- 'digest' is a checksum of the payload of the request,
- 'host' is the hostname of the recipient of the request (remote when signing outgoing request, local on incoming request),
- 'Signature' contains the signature generated using the private key and details on its generation:

>   * 'keyId' is a unique id, formatted as an url. hostname is used to retrieve the public key via custom discovery

Reviewer: What does that mean? If the mentioned discovery is querying the OCM `.well-known` endpoint, then this may not be well defined if the hostname is different.

Reviewer 3: https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12#section-2.1.1
Maybe we should say in our case keyId should be an FQDN? And then we use .well-known with fallback to DNS SRV from that FQDN.

Reviewer 3: Let's stick to URL as proposed here, we can discuss in a follow-up whether FQDN is more appropriate for us.

  * 'algorithm' specify the algorithm used to generate signature
  * 'headers' specify the properties used when generating the signature
  * 'signature' the signature of an array containing the properties listed in 'headers'. Some properties like content-length, date, digest, and host are mandatory to protect against authenticity override.

### How to generate the Signature for outgoing request

After properties are set in the headers, the Signature is generated and added to the list.  

This is a quick PHP example of headers for outgoing request:

```php
    $headers = [
        '(request-target)' => 'post /path',
        'content-length' => strlen($payload),
        'date' => gmdate('D, d M Y H:i:s T'),
        'digest': 'SHA-256=' . base64_encode(hash('sha256', utf8_encode($payload), true)),
        'host': 'hostname.of.the.recipient',
    ];

    openssl_sign(implode("\n", $headers), $signed, $privateKey, OPENSSL_ALGO_SHA256);
		
    $signature = [
        'keyId' => 'https://author.hostname/key',

>         'algorithm' => 'ras-sha256',

Reviewer 3: rsa-sha256

        'headers' => 'content-length date digest host',
        'signature' => $signed 
    ];

    $headers['Signature'] = implode(',', $signature);
```

### How to confirm Signature on incoming request

The first step would be to confirm the validity of each properties:

- '(request-target)' and 'host' are immutable to the type of the request and the local/current host,
- 'content-length' and 'digest' can be re-generated and compared from the payload of the request,
- A maximum TTL must be applied to 'date' and current timestamp, 
- regarding data extracted from 'Signature':
  * using 'keyId' to get the public key from remote signatory,
  * 'headers' is used to generate the clear version of the signature and must contain at least 'content-length', 'date', 'digest' and 'host',
  * 'signature' is the encrypted version of the signature.

Using the clear version of the signature, the public key and the encrypted version, it is easy to confirm the authenticity of the request:

> ```php

Reviewer: I'd suggest to describe this with an abstract pseudo-language, as opposed to PHP and its PHP-specific libraries.

    $clear = [
        '(request-target)' => 'post /path',
        'content-length' => strlen($payload),
        'date' => 'Mon, 08 Jul 2024 14:16:20 GMT',
        'digest': 'SHA-256=' . base64_encode(hash('sha256', utf8_encode($payload), true)),
        'host': $localhost
    ];
    
    $signed = "DzN12OCS1rsA[...]o0VmxjQooRo6HHabg==";
    if (openssl_verify(implode("\n", $clear), $signed, $publicKey, 'sha256') !== 1) {
        throw new InvalidSignatureException('signature issue');
    }
```

### Validating the payload

> while signed, it is still needed to confirm the validity of the payload.  
> The last step is to ensure that the payload implies action initiated on the behalf the source of the request.  

Reviewer: ```suggestion
Following the validation of the signature, the host should also confirm the validity of the payload, that is ensuring that the actions implied in the payload actually initiated on behalf of the source of the request.
```

> As an example, if the payload is about initiating a new share the file owner has to be an account from the instance at the origin of the request.  

Reviewer: Apart the minor rephrasing I proposed above, I'm not sure this can actually be done in general, and the example in particular does not seem feasible: how can an EFSS validate something that refers to user belonging to a remote EFSS? Ultimately it has to trust the remote EFSS and its payload.

Reviewer 3: Sure, but I think what is meant is: if the owner is specified as `alice@cernbox.cern.ch` but the sender is `cloud.pondersource.com` then the receiving server should flag the domain name mismatch.

We could also make this automatic in a future version of the spec, so that owner would just be `alice` and the `@cernbox.cern.ch` would be implied silently from the keyId of the signature.
